### PR TITLE
plutus-ledger: implement the separate MPS proposal

### PR DESCRIPTION
**Pull request by michaelpj**
This changes the rules for currency forging.
- A new field is added to `Tx` which contains the monetary policies
that will be run against the transaction.
    - These have a different signature: `PendingTxMPS -> ()`
    - They get a different kind of `PendingTx`, because the "current
item" is different.
- Validation runs all the monetary policies, they must pass.
- For every currency forged, the corresponding monetary policy must
appear in the transaction.

This has some knock-on effects:
- The `Currency` example becomes simpler, because we can skip one of
the transactions.
- The `GameStateMachine` uses the "forwarding MPS" technique, whereby
the MPS just asserts that the normal validator runs. This requires
a circular dependency in their hashes, which we solve by having the
MPS hash carried in the machine state.

---

The `GameStateMachine` example could be changed to not require the additional "initialized" state and transition, but then it wouldn't be using a forwarding MPS any more (the MPS would need logic to decide whether the state was initial). I opted not to change that for now, but I could do.

### CODE REVIEW

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs [hunk 1]
```diff
@@ -175,6 +175,13 @@ scriptInstance = Scripts.validator @(SM.StateMachine GameState GameInput)
     where
         wrap = Scripts.wrapValidator @GameState @GameInput
 
+{-# INLINABLE monetaryPolicy #-}
```
Comment (j-mueller) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs]: Does this need to be inlinable?

File: plutus-wallet-api/ledger/Ledger/Index.hs [hunk 2]
```diff
@@ -109,7 +109,7 @@ data ValidationError =
     -- ^ The current slot is not covered by the transaction's validity slot range.
     | SignatureMissing PubKey
     -- ^ The transaction is missing a signature
-    | ForgeWithoutScript Scripts.ValidatorHash
+    | ForgeWithoutScript Scripts.MonetaryPolicyHash
     -- ^ The transaction attempts to forge value of a currency without spending
```
Comment (j-mueller) [inline plutus-wallet-api/ledger/Ledger/Index.hs]:
```suggestion
    -- ^ The transaction attempts to forge value of a currency without running
```

File: plutus-wallet-api/ledger/Ledger/Index.hs [hunk 3]
```diff
@@ -109,7 +109,7 @@ data ValidationError =
...
     --   a script output from the address of the currency's monetary policy.
```
Comment (j-mueller) [inline plutus-wallet-api/ledger/Ledger/Index.hs]:
```suggestion
     --   the currency's monetary policy script.
```

File: plutus-wallet-api/ledger/Ledger/Scripts.hs [hunk 4]
```diff
@@ -306,6 +351,17 @@ runScript checking (ValidationData valData) (Validator validator) (DataValue dat
     let appliedValidator = ((validator `applyScript` (fromCompiledCode $ liftCode dataValue)) `applyScript` (fromCompiledCode $ liftCode redeemer)) `applyScript` (fromCompiledCode $ liftCode valData)
     evaluateScript checking appliedValidator
 
+-- | Evaluate a monetary policy script just the validation data, returning the log and a boolean indicating whether evaluation was successful.
```
Comment (j-mueller) [inline plutus-wallet-api/ledger/Ledger/Scripts.hs]: Doesn't return a boolean

**Comment by michaelpj** [comment]
Also, this doesn't do redeemers for MPSs, which I think we wanted. Should think about adding that.

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs]: nope

**Pull request merged**